Move expr_cast from refinement to util/convert_expr.h

romainbrenguier opened the pull request:
This moves these utility functions to util which is more appropriate.
The file and functions are also renamed to avoid confusion with util/expr_cast.h

romainbrenguier committed:
Remove solvers/refinement/expr_cast.h

numeric_cast can now be used instead.

tautschnig commented on src/util/convert_expr.h:
```diff
@@ -9,20 +9,19 @@ Author: Diffblue ltd. All Rights Reserved
-/// TODO: this should go to util
 template<typename T>
-struct expr_cast_implt final { };
+struct from_expr_implt final { };
```
Apparently this is more than just a plain file renaming? It's also a very confusing one, as there already are functions called `from_expr`.

romainbrenguier replied:
`from_expr` functions convert an expression to something else (which is also the case here). Here the type we are converting to is given as template argument so the meaning should be clear. But I'm open to suggestions for better names.

tautschnig replied:
To be honest, I'm not very happy with the "convert" terminology. You are doing (partly) checked type casts, nothing more and nothing less. To me, "convert" suggests that there is some sort of transformation going on (which would be true of the existing `from_expr` code).

tautschnig replied:
In that vein: `checked_cast_...` maybe? Also call the file `checked_expr_cast.h`?

romainbrenguier replied:
Checked typecast is more what is done in expr_cast.h and it is a bit different from these functions.
In particular `expr_checked_cast` is already used there.

tautschnig replied:
Honestly, I have no idea what this code is doing. The file used to be named "expr_cast.h", hence I made my inferences. There is a single comment in this file and having read that comment block three times now I still have no clue what it is trying to tell me (what is a "specific type", what is "empty" about an "optional" and how does an "optional" come into play when we are talking about `exprt`?). You are now saying it's "a bit different"? How much is "a bit"?

LAJW replied:
(I wrote the original version). I wanted to follow the naming convention of `boost::lexical_cast`, only rather than throwing exceptions, I wanted to return empty optional to signal "bad conversion". It was developed independently of `expr_dynamic_cast`. But then the `expr_dynamic_cast` was merged (after `expr_cast` was already used locally in `string_refinement`). (`expr_dynamic_cast` and `expr_try_dynamic_cast` return references and pointers, but only of types derived from `exprt`).

Personally, I'd leave it as "expr_cast" and let it be the function which converts all sorts of `exprts` to various types in optional, and returning empty optional on error.

expr_cast.h doesn't actually have a function called "expr_cast", so it could be declared there.

tautschnig replied:
You are making it sound like there are multiple somewhat competing implementations? That's interesting.

Maybe it would help (this might be what you are suggesting in your last sentence) to have all of this in one file? Then at least there would be more context to help understand the code.

LAJW replied:
There is a single implementation for casting between `exprt`s with two different interfaces - throw on error or return `nullptr` on error (in `util/expr_cast.h`). This function however casts from `exprt` to an integer, string, or whatever other types we can reliably cast to. But since `exprt` is cheap to copy, it could also use the same pattern - `return optionalt<derived_exprt>`. Why it wasn't used in the first place - I have no idea.

tautschnig replied:
Thank you very much for providing further detail -- what does it even mean to "cast" an `exprt` in this way? Anyway at least I believe I understand what the code is supposed to do. I have no idea how this code ever made it into the code base in its current form, but then that's a different question that the approvers of #1241 need to answer.

Anyway, it seems `util/arith_tools.h` seems to be the place for most of them.

LAJW replied:
I don't think that conversion from `exprt` to string or from one `exprt` to another is an arithmetic operation.

It made its way into the code, as it provides saner way of doing those things compared to existing 15+ line dance with tons of `INVARIANTS` every time someone wanted to cast const_exprt to its C++ representation. The only "problem" is the name, which is debatable.

tautschnig replied:
> I don't think that conversion from exprt to string or from one exprt to another is an arithmetic operation.

Indeed, I generalised too quickly - the integer conversion ones should go in `arith_tools.h`, `exprt` to `*_exprt` should go in `expr_cast.h`, and `exprt` to string I have no idea what that is.

> The only "problem" is the name, which is debatable.

The problems certainly include the complete absence of documentation as well as poor choice of placement in the file system. Is it saner? I don't know, because I don't know what it's supposed to do (as there is no documentation).

smowton replied:
Re: `expr_dynamic_cast` and returning pointers vs. `optionalt`, I believe @reuk was the person advocating pointers because the syntax to use them is briefer -- `optionalt<reference_wrapper<T>>` requires an explicit `.get()` operation a little too often.

LAJW replied:
@smowton Before expr_dynamic_cast, there was this, which was intended to return `optional<T>` instead of `optional<reference_wrapper<T>>`. Old casts (`to_****_expr`) were doing that, because it's not inefficient, so I followed, as it was the safest and most concise way of doing those things.

tautschnig reviewed (dismissed):
This is a very partial review: introducing `from_expr` with a very different meaning than the pre-existing `from_expr` API is confusing.

smowton reviewed (changes requested):
I think the two casts we can see here are different enough to deserve different names. The cast to `refined_string_exprt` should be a case of `expr_dynamic_cast` and implemented as such, whereas the to-integer conversions are helpers on top of `util/arith_tools.h`. I suggest:

1. Implement `can_cast_expr` and `validate_expr` in a string-refinement header (alongside the definition of `refined_string_exprt`) so that you can use `expr_dynamic_cast` with it.

2. Move the constant-expr-to-integer stuff into `arith_tools.h`. Deprecate the existing `to_integer` and `to_unsigned_integer` (but don't remove them) and replace with your new optional-returning or throwing versions.

romainbrenguier committed:
Add validate and can_cast method to string_exprt

romainbrenguier commented:
@smowton Thanks for the suggestion. I didn't deprecate `to_unsigned_integer` yet because for now there is no `from_expr<unsigned>` implemented, but it would be easy to add.

smowton commented:
Hmmm, two more nitpicks:

1. I think you should take `constant_exprt` rather than `exprt`, even though that's a bit more verbose at certain use-sites. It distinguishes failure because this was the wrong kind of exprt vs. failure because the constant wasn't numeric / was too large / whatever.

2. std::size_t's range may differ from `long` (as in `to_long()`), how about using `std::numeric_limits<...>.max()` etc to check in range, and thus generalise the `size_t` case into any integer type?

tautschnig commented:
> std::size_t's range may differ from long (as in to_long()), how about using std::numeric_limits<...>.max() etc to check in range, and thus generalise the size_t case into any integer type?

There's `integer2size_t` covering this, and that's really what should be used here.

smowton commented:
Whichever function does the conversion, I'd suggest templating on a numeric type rather than manually enumerating integer2long, integer2size_t, integer2uint16_t, ...

tautschnig commented:
> Whichever function does the conversion, I'd suggest templating on a numeric type rather than manually enumerating integer2long, integer2size_t, integer2uint16_t, ...

True, but either way it should only be done in one place. If someone wants to step up with a PR doing the refactoring that's fine with me.

romainbrenguier commented:
@smowton @tautschnig 
> Whichever function does the conversion, I'd suggest templating on a numeric type rather than manually enumerating integer2long, integer2size_t, integer2uint16_t, ...

Not sure that it is exactly what you had in mind but I implemented some conversion template which should help with that: https://github.com/diffblue/cbmc/pull/1582

allredj commented:
Shall I review this or is it still likely to change?

romainbrenguier commented:
@allredj please wait for https://github.com/diffblue/cbmc/pull/1582 to be merged, I will then adapt this one

romainbrenguier committed:
Use numeric_cast instead of refinement/expr_cast

romainbrenguier committed:
Extend numeric_cast for constant expressions

This requires moving some declarations to arith_tools, because overload
needs to be declared in the same struct.

This also adds a numeric_cast_v function which returns a value instead
of an optional but an invariant can fail.

tautschnig commented:
It seems #1582 has been merged, so it may be possible to continue on this one?

romainbrenguier commented:
@tautschnig yes but at the moment I'm blocked by the visual studio compilation problem. If someone has an idea?

tautschnig commented:
I believe the VS failure is due to https://connect.microsoft.com/VisualStudio/feedback/details/1211985. You'll need to make those runtime checks for MSVC, I'm afraid.

smowton commented:
Gah, https://connect.microsoft.com/VisualStudio/feedback/details/1211985

As @tautschnig says. https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B suggests `!defined(_MSC_VER) || _MSC_VER >= 1900` should diagnose a compiler with a properly `constexpr` `max()` method.

romainbrenguier committed:
Use numeric_cast instead of other conversion

This is to demonstrate the use of numeric_cast

romainbrenguier commented:
@tautschnig @smowton @allredj this is now working with VS and ready for rereview

tautschnig commented on src/util/mp_arith.h:
```diff
@@ -105,28 +114,53 @@ struct numeric_castt<T,
+  return numeric_castt<Target, Source>::numeric_cast(arg);
+}
+
+/// An invariant with fail with message "Bad conversion" if conversion
```
s/with/will/? Also, the first sentence should say what it does, not just when it fails.

tautschnig commented on src/util/mp_arith.h:
```diff
@@ -105,28 +114,53 @@ struct numeric_castt<T,
+/// \param arg: constant expression
+/// \return value of type T
+template <typename Target, typename Source>
+Target numeric_cast_v(const Source &arg)
```
What does `_v` stand for?

tautschnig commented on src/util/string_expr.h:
```diff
@@ -241,4 +241,18 @@ inline const refined_string_exprt &to_string_expr(const exprt &expr)
+  INVARIANT(x.id() == ID_struct, "refined string expr are struct");
+  INVARIANT(
+    x.operands().size() == 2,
+    "refined string expr have length and content fields");
```
Use `validate_operands`?

tautschnig commented on src/solvers/refinement/string_constraint_generator_transformation.cpp:
```diff
@@ -442,8 +442,8 @@ static optionalt<std::pair<exprt, exprt>> to_char_pair(
   const auto expr2_str = get_string_expr(expr2);
-  const auto expr1_length=expr_cast<size_t>(expr1_str.length());
-  const auto expr2_length=expr_cast<size_t>(expr2_str.length());
+  const auto expr1_length = numeric_cast<size_t>(expr1_str.length());
```
Why does this one work without a second template argument?

tautschnig commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -795,12 +794,13 @@ static optionalt<exprt> get_array(
 
-  unsigned n;
-  if(to_unsigned_integer(to_constant_expr(size_val), n))
+  auto n_opt = numeric_cast<unsigned, exprt>(size_val);
```
The counterpart of the above comment: why does this require a second template argument?

tautschnig reviewed (commented):
I'm confused by the sometimes-use-a-second-template-argument, and if the second template argument (which always is `exprt`) is truly required then this is not at all beautiful.

smowton commented:
Given that the possible source types are exactly `exprt` and `constant_exprt`, would it be simplest to omit the source template parameter and just introduce overloads for the two cases, one of which hard-casts the `exprt` to `constant_exprt` and calls the other?

romainbrenguier commented:
@smowton 
> Given that the possible source types are exactly exprt and constant_exprt, would it be simplest to omit the source template parameter and just introduce overloads for the two cases, one of which hard-casts the exprt to constant_exprt and calls the other?

Not sure I understood this correctly, but `mp_integer` is also a possible source type.

smowton commented:
As in you can `numeric_cast` from `mp_integert` to `size_t` for example? Can you also `numeric_cast` between primitives?

romainbrenguier commented:
@smowton 
> As in you can numeric_cast from mp_integert to size_t for example? 

Yes you can (and to any integral type)

> Can you also numeric_cast between primitives?

Not at the moment but it could be added if someone finds it useful.

smowton commented:
OK, understood. In this case removing the source template arg is beyond my skill level, but @reuk or @LAJW can probably work magic if asked nicely :)

reuk commented:
I'm pretty sure that's possible, yeah. I think you just need an overload of `numeric_cast` in the struct for each possible `Source` type.

reuk commented:
Having said that, this does require moving the `exprt` overload back into the declaration in mp_arith.h, which I think is what you're trying to avoid...

reuk commented:
Created a [pr showing how this might look](https://github.com/romainbrenguier/cbmc/pull/1)

allredj commented on src/util/string_expr.h:
```diff
@@ -241,4 +241,18 @@ inline const refined_string_exprt &to_string_expr(const exprt &expr)
+template <>
+inline bool can_cast_expr<refined_string_exprt>(const exprt &base)
+{
+  return base.id() == ID_struct && base.operands().size() == 2;
```
Shouldn't we check anything about the type of operands?

allredj commented on src/util/arith_tools.h:
```diff
@@ -11,21 +11,170 @@ Author: Daniel Kroening, kroening@kroening.com
+      std::numeric_limits<T>::min() >=
+      std::numeric_limits<decltype(mpi.to_long())>::min(),
+      "Numeric cast only works for types smaller than long long");
+#else
```
Could you please add the same comment as below?

allredj commented on src/util/arith_tools.h:
```diff
@@ -11,21 +11,170 @@ Author: Daniel Kroening, kroening@kroening.com
+///         empty optional otherwise.
+template <typename T>
+struct numeric_castt<T,
+  typename std::enable_if<std::is_integral<T>::value &&
```
This seems to be the same function as above, except for the `mpi` function called. Could we refactor?

romainbrenguier replied:
I refactored a bit to have only one structure for the conversion to integral type, but I still have one function for mpi to signed and one function from mpi to unsigned because all the checks are different.

reuk replied:
Are you sure? I'm pretty sure the checks can be made the same (I show how this is done in my proposal PR)

romainbrenguier replied:
Yes it can be done, I will try that.

allredj commented on src/util/arith_tools.h:
```diff
@@ -11,21 +11,170 @@ Author: Daniel Kroening, kroening@kroening.com
+/// \param arg: expression to convert
+/// \return optional integer of type T if conversion is possible,
+///         empty optional otherwise.
+template <typename Target>
```
Why do we call that `Target` here and not T? Update the doc header.

romainbrenguier replied:
this is to make it clear this is the type we convert to, and not from.

allredj commented on src/util/arith_tools.h:
```diff
@@ -11,21 +11,170 @@ Author: Daniel Kroening, kroening@kroening.com
+  return *maybe;
+}
+
+/// Convert anexpression to integral type Target
```
`anexpression` -> `an expression`

allredj commented on src/util/arith_tools.h:
```diff
@@ -11,21 +11,170 @@ Author: Daniel Kroening, kroening@kroening.com
+/// \param arg: mp_integer
+/// \return value of type Target
+template <typename Target>
+Target numeric_cast_v(const mp_integer &arg)
```
What does `_v` mean?

romainbrenguier replied:
It stands for value (this is to distinguish from optionalt). This was the notation used in the `expr_cast` I removed, but I'm open to better suggestions.

allredj replied:
`_val`?

tautschnig commented on src/solvers/refinement/string_constraint_generator_transformation.cpp:
```diff
@@ -442,8 +442,8 @@ static optionalt<std::pair<exprt, exprt>> to_char_pair(
-  const auto expr1_length=expr_cast<size_t>(expr1_str.length());
-  const auto expr2_length=expr_cast<size_t>(expr2_str.length());
+  const auto expr1_length = numeric_cast<size_t>(expr1_str.length());
+  const auto expr2_length = numeric_cast<size_t>(expr2_str.length());
```
`std::size_t` for consistency. (Also applies above.)

peterschrammel commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -852,13 +852,11 @@ static optionalt<exprt> get_array(
   exprt size_expr=to_array_type(arr.type()).size();
-  PRECONDITION(size_expr.id()==ID_constant);
-  to_unsigned_integer(to_constant_expr(size_expr), n);
+  auto n = numeric_cast_v<unsigned>(size_expr);
```
Why do you use sometimes unsigned and sometimes sizet for values that denote sizes?

romainbrenguier committed:
Change length argument type to size_t

romainbrenguier committed:
Change type of size to std::size_t

peterschrammel commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -825,9 +825,9 @@ static optionalt<exprt> get_array(
-      unsigned idx;
-      if(!to_unsigned_integer(to_constant_expr(index), idx) && idx<n)
-        initial_map[idx] = arr_val.operands()[i + 1];
+      if(auto idx = numeric_cast<std::size_t>(index))
```
I'd appreciate braces here.

peterschrammel reviewed (approved):
Lgtm
